### PR TITLE
[MSHARED-1019] Allow pass raw cli option to Maven process

### PR DESCRIPTION
**Pull request by slawekjaranowski**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/MSHARED) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MSHARED-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSHARED-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Review by olamy — APPROVED** [review]

**Comment by michael-o** [comment]
I do not fully understand how this works. Can I set args one by one as with a list/array?

**Comment by slawekjaranowski** [comment]
You provide whole additional args as String.

argLine is passed to shared-utils - `Commandline#createArg()` and finally is parsed by `CommandLineUtils.translateCommandline()`

**Comment by michael-o** [comment]
What a brittle pain! This will fail sooner or later. I rather prefer an arrray.

**Comment by slawekjaranowski** [comment]
😄 
No problem we can use array/list of String for argLine in `InvocationRequest`
next each element  will be passed as-is to `Argument#setValue`

https://github.com/apache/maven-shared-utils/blob/master/src/main/java/org/apache/maven/shared/utils/cli/Commandline.java#L476-L492

 Will be better?

**Comment by michael-o** [comment]
> 😄 No problem we can use array/list of String for argLine in `InvocationRequest` next each element will be passed as-is to `Argument#setValue`
> 
> https://github.com/apache/maven-shared-utils/blob/master/src/main/java/org/apache/maven/shared/utils/cli/Commandline.java#L476-L492
> 
> Will be better?

Sorry, I seem to be blind again. Which do you compare? A per-line approach or a parser?!

Basically, what I don't understand why we need to parse here although the MavenCli class will do that. Talked about this several times with @cstamas .

**Comment by slawekjaranowski** [comment]
Now I'm also confused .... 
Do you think about `InvocationRequest` which contains method like .... `setDebug`, `setOffline` ... and so on for other Maven options?, So it is historical ... and for compatibility must be preserved 

With my PR you needn't us it, and simply:

```
invocationRequest.setArgLineAtEnd("-e -B clean verify");
invoker.execute( invocationRequest );
```

**Comment by olamy** [comment]
uhm I tend to agree parsing could be a pain for something such 

```
invocationRequest.setArgLineAtEnd("-e -B clean verify -Dfile=c:\\My Files\grhhh windows");
```

**Comment by slawekjaranowski** [comment]
I don't parse argLinie, it is pass **as-is** to shared-utils - `Commandline#createArg().setLine(...)`, 
what happens in shared-utils is hidden for invoker

**Comment by michael-o** [comment]
@slawekjaranowski Yes, retain previous ones is correct, but new ones should be:

```
invocationRequest.addArg("-B")
invocationRequest.addArg("-Dfoo=bar")
```

I have a very special opinion CLI tools on shared utils. I consider 80% of the code as useless.

File: src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java [hunk 1]
```diff
@@ -104,6 +104,8 @@ public Commandline build( InvocationRequest request )
 
         setThreads( request, cli );
 
+        setArgsLine( request, cli );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java:107]: Wouldn't setArgs be better?

Comment (slawekjaranowski) [reply src/main/java/org/apache/maven/shared/invoker/MavenCommandLineBuilder.java]: fixed

**Review by michael-o — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #143

**Pull request merged**